third_party/2and3/cryptography/hazmat/primitives/asymmetric/utils.pyi

```diff
@@ -1,6 +1,10 @@
-from typing import Tuple
+from typing import Optional, Tuple
+from cryptography.hazmat.primitives.hashes import HashAlgorithm
 
 def decode_dss_signature(signature: bytes) -> Tuple[int, int]: ...
 def encode_dss_signature(r: int, s: int) -> bytes: ...
 
-class Prehashed(object): ...
+class Prehashed(object):
+    _algorithm: HashAlgorithm
+    _digest_size: int
+    def __init__(self, algorithm: Optional[HashAlgorithm]) -> None: ...
```

Comment:
Why is this Optional? The code you link throws an error if it's not a HashAlgorithm.

Reply:
There are two valid paths through the code. One for algorithm being a HashAlgorithm and one if it is not.

Reply:
Not according to the code you linked: https://cryptography.io/en/latest/_modules/cryptography/hazmat/primitives/asymmetric/utils/#Prehashed

Reply:
To my understanding, one way returns you a `Prehashed` and the other one a `TypeError`. `mypy` is smart enough to drop the `Optional` after the `isinstance` check and having the `Optional` makes life easier, when `algorithm` can be `None` (e.g. if set at runtime and) and the developer expects to handle this through the `TypeError`.

Reply:
Part of the point of typechecking is to tell users beforehand that something they pass to a function will throw an error.

Reply:
Ok, the documentation does not mention the TypeError, so that behaviour is not guaranteed.
i stand corrected and removed the `Optional`.